src/docfx/build/learn/LearnHierarchyBuilder.cs

```diff
@@ -0,0 +1,123 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.Docs.Validation;
+using Newtonsoft.Json.Linq;
+
+namespace Microsoft.Docs.Build
+{
+    internal class LearnHierarchyBuilder
+    {
+        private readonly ErrorBuilder _errors;
+        private readonly ContentValidator _contentValidator;
+
+        private readonly ListBuilder<LearningPath> _learningPaths = new ListBuilder<LearningPath>();
+        private readonly ListBuilder<Module> _modules = new ListBuilder<Module>();
+        private readonly ListBuilder<ModuleUnit> _moduleUnits = new ListBuilder<ModuleUnit>();
+        private readonly ListBuilder<Achievement> _achievements = new ListBuilder<Achievement>();
+
+        public LearnHierarchyBuilder(ErrorBuilder errors, ContentValidator contentValidator)
+        {
+            _errors = errors;
+            _contentValidator = contentValidator;
+        }
+
+        public void ValidateHierarchy()
+        {
+            _contentValidator.ValidateHierarchy(GetAllLearnHierarchyModels().ToList());
+        }
+
+        public void AddLearningPath(JObject content)
+        {
+            var path = JsonUtility.ToObject<LearningPath>(_errors, content);
+            _learningPaths.Add(path);
+
+            if (path.Trophy != null && GetAchievement(path.Trophy, out var achievement) && achievement != null)
```

Comment:
Ensure this method is idempotent, that is, the method should merely add the item to the collection. Move these achievement logic to the `ValidateHierarchy` method.

Reply:
OK